torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -129,15 +134,192 @@ def get_tp_parallel_strategy(
+    model, world_mesh, parallel_dims, job_config: JobConfig, device, model_config: Dict
+):
+    """
+    This API gets individual torch.nn.Module objects for each pipeline stage (including virtual stages).
+
+    The SPMD parallelisms should be applied to
+    """
+    pp_mesh = world_mesh["pp"]
+    pp_rank = pp_mesh.get_local_rank()
+    pp_size = pp_mesh.size()
+    microbatches = (
+        job_config.experimental.pipeline_parallel_microbatches or parallel_dims.pp
+    )
+    stage_idx = pp_rank
+    this_stage_layer_names = split_stage_fqns(
+        _llama_fqns(len(model.layers)),
+        job_config.experimental.pipeline_parallel_split_points,
+        pp_rank,
+    )
+
+    if pp_rank < pp_size - 1:
+        model.norm = None
+        model.output = None
+    if pp_rank > 0:
+        model.tok_embeddings = None
+    names = list(model.layers.keys())
+    for name in names:
+        if f"layers.{name}" not in this_stage_layer_names:
+            del model.layers[name]
+
+    logger.info(f"PP rank {pp_rank} is using this model chunk\n{model}")
+
+    # TODO(whc) once ManualPipelineStage supports lazy shape inference, we can leave model on meta device longer and
+    # get rid of the input shape hardcoded here. For now, it should not be a big deal since we only materialize the
+    # layers of the model that map to this stage, not the whole model.
+
+    if pp_rank == 0:
+        # first layer
+        input = torch.randint(
+            model_config.vocab_size,
+            size=(job_config.training.batch_size, job_config.training.seq_len),
+            dtype=torch.int64,
+            device=device,
+        )
+    else:
+        # later layers (assume all start w/ a transformer layer)
+        input = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param]
+            if parallel_dims.dp_enabled
+            else torch.float32,
+            device=device,
+        )
+
+    if pp_rank == pp_size - 1:
+        # last layer
+        output = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.vocab_size,
+            ),
+            dtype=torch.float32,
+            device=device,
+        )
+    else:
+        # earlier layers (assume all end in a transformer layer)
+        output = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param]
+            if parallel_dims.dp_enabled
+            else torch.float32,
+            device=device,
+        )
```

Comment:
i think those lines are llama-model specific? thats why i kept them here.  (The plan is to get rid of these still by providing a runtime shape inference thing, but it might not be the highest priority).